clojask-script/doc/parameters.md
{:title "Parameters"
 :date "2021-01-23"
 :layout :post
 :tags  []
 :toc true}


<style>
/* table styles */
table, th, td {
  border: 1px solid black;
  padding: 5px;
}
td {
  padding: 10px;
}
</style>

<br>

**`ns: clojure-backtesting.parameters`**

This namespace defines the parameters that are configurable in the backtester.

---

## Trading Price

The trading price is by default set to be closing price.

You can change it to opening price by changing the parameter.clj in `src/`:

```clojure
(def PRICE-KEY :OPRNPRC)
```

-----------------

## Margin Requirements

<br>

### `INITIAL-MARGIN`

This variable stores the minimum ratio of cash to total value of securities that must be paid when buying on margin. The default value is set as `0.5`.

**Example:**

```clojure
;; check the variable
(println INITIAL-MARGIN)

;; output:
0.5
```

<br>

### `update-initial-margin`

This function updates the initial margin.

**Parameters:**

- `num` - new initial margin value, needs to be > 0

**Example:**

```clojure
;; update the variable
(update-initial-margin 0.7)
(println INITIAL-MARGIN)

;; output:
0.7
```

<br>

### `MAINTENANCE-MARGIN`

This variable stores the minimum ratio of cash to total value of securities that must be maintained in the margin account. If the portfolio margin goes below the maintenance margin, all positions will be automatically closed. The default value is set as `0.25`.

**Example:**

```clojure
;; check the variable
(println MAINTENANCE-MARGIN)

;; output:
0.25
```

<br>

### `update-maintenance-margin`

This function updates the maintenance margin.

**Parameters:**

- `num` - new maintenance margin value, needs to be > 0

**Example:**

```clojure
;; update the variable
(update-maintenance-margin 0.2)
(println MAINTENANCE-MARGIN)

;; output:
0.2
```

---

## Transaction Cost & Loan Interest

<br>

### `TRANSACTION-COST`

This variable stores the rate of the commission fee that is charged upon the execution of an order. The default value is `0.0`.

**Example:**

```clojure
;; check the variable
(println TRANSACTION-COST)

;; output:
0.0
```


<br>

### `update-transaction-cost`

This function updates the transaction cost.

**Parameters:**

- `num` - new initial margin value, needs to be within the range of [0,1).

**Example:**

```clojure
;; update the variable
(update-transaction-cost 0.2)
(println TRANSACTION-COST)

;; output:
0.2
```

<br>


### `INTEREST-RATE`

This variable stores the simple interest rate (per annum) that is incurred when making a loan. Loan interests will be deducted at the end of every month. The default value is `0.0`.

**Example:**

```clojure
;; check the variable
(println INTEREST-RATE)

;; output:
0.0
```

<br>

### `update-interest-rate`

This function updates the interest rate.

**Parameters:**

- `num` - new initial margin value, needs to be within the range of [0,1).


```clojure
;; update the variable
(update-interest-rate 0.3)
(println INTEREST-RATE)

;; output:
0.3
```

<br>


clojask-script/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-10-06
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-10-06
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojask-script/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojask-script/compare/0.1.0...0.1.1


src/clojure_backtesting/parameters.clj
(ns clojure-backtesting.parameters)

;; ============ CONFIGURABLE PARAMETERS ============

(def PRINT false)
(def DIRECT true)
(def MAXLOOKAHEAD 10) ; This parameter is used in `get-pre-date-and-content`
(def ORDER-EXPIRATION 3)
(def LEVERAGE true)
(def MAXDISCONTINUITY 10)
(def MAXRANGE 50000)
(def MAXMERGERANGE "1 year")
(def PRICE-KEY :PRC) ;; trade at closing price
;; (def PRICE-KEY :OPENPRC) uncomment this if you want to trade at opening price instead

;; ============== Cache =================

(def CACHE-SIZE 30)
(defn CHANGE-CACHE-SIZE
  [size]
  (def CACHE-SIZE size))

;; =============== Indicator ===============
(def ^:dynamic EMA-CYCLE 20)
(def MACD-SIGNAL 9)
(def MACD-SHORT 12)
(def MACD-LONG 26)
(def ^:dynamic RSI-CYCLE 14)

(defn CHANGE-EMA-CYCLE
  [int]
  (def ^:dynamic EMA-CYCLE int)
  (def ^:dynamic EMA-K (/ 2 (+ EMA-CYCLE 1)))
  )

(defn CHANGE-MACD-SIGNAL
  [int]
  (def MACD-SIGNAL int)
  (def MACD-SIGNAL-K (/ 2 (+ MACD-SIGNAL 1))))

(defn CHANGE-MACD-SHORT
  [int]
  (def MACD-SHORT int)
  (def MACD-SHORT-K (/ 2 (+ MACD-SHORT 1))))

(defn CHANGE-MACD-LONG
  [int]
  (def MACD-LONG int)
  (def MACD-LONG-K (/ 2 (+ MACD-LONG 1))))

(defn CHANGE-RSI-CYCLE
  [int]
  (def ^:dynamic RSI-CYCLE int))

;; ============ Parameters for margin requirements ============

(def INITIAL-MARGIN 0.5) 
;; This is the min cash / order total 
;; Set it to nil to enable infinite margin

(def MAINTENANCE-MARGIN 0.25)
;; When portfolio margin is < maintenance margin, all positions will be closed


;; ============ Parameters for interests & transaction costs ============

(def TRANSACTION-COST 0.0)
;; Commission fee (%) that needs to be paid when making a purchase, must be < 1.0

(def INTEREST-RATE 0.0)
;; The simple interest rate (p.a.) for making a loan
;; Usually between 3% and 12% per annum



;; ============ Functions for updating parameters ============

(defn update-initial-margin 
    "This function updates the initial margin."
	[new-im]
    (if (or (> new-im 0) (= new-im nil))
        (def INITIAL-MARGIN new-im)
        (println "Failed: The initial margin needs to be greater than zero.")))

(defn update-maintenance-margin 
    "This function updates the maintenance margin."
	[new-mm]
    (if (> new-mm 0)
        (def MAINTENANCE-MARGIN new-mm)
        (println "Failed: The maintenance margin needs to be greater than zero.")))

(defn update-interest-rate
    "This function updates the interest rate."
	[new-ir]
    (if (and (pos? new-ir) (< new-ir 1))
        (def INTEREST-RATE new-ir)
        (println "Failed: The interest rate needs to be within the range of [0,1).")))

(defn update-transaction-cost
    "This function updates the transaction cost."
	[new-tc]
    (if (and (pos? new-tc) (< new-tc 1))
        (def TRANSACTION-COST new-tc)
        (println "Failed: The transaction cost needs to be within the range of [0,1).")))

;; ============ FIXED PARAMETERS ============

;; (def NOMATCH "2oif94ksdajf09934")
(def ^:dynamic EMA-K (/ 2 (+ EMA-CYCLE 1)))
(def MACD-SIGNAL-K (/ 2 (+ MACD-SIGNAL 1)))
(def MACD-SHORT-K (/ 2 (+ MACD-SHORT 1)))
(def MACD-LONG-K (/ 2 (+ MACD-LONG 1)))

(def TICKER-KEY :PERMNO)